Framework/MDAlgorithms/src/IntegrateEllipsoidsTwoStep.cpp

```diff
@@ -0,0 +1,553 @@
+  for (size_t i = 0; i < n_peaks; i++) // Note: we skip un-indexed peaks
+  {
+    V3D hkl(peaks[i].getH(), peaks[i].getK(), peaks[i].getL());
+    if (Geometry::IndexingUtils::ValidIndex(hkl, 1.0)) // use tolerance == 1 to
+                                                       // just check for (0,0,0)
+    {
+      peak_q_list.emplace_back(peaks[i].getQLabFrame());
+      V3D miller_ind(static_cast<double>(boost::math::iround<double>(hkl[0])),
+                     static_cast<double>(boost::math::iround<double>(hkl[1])),
+                     static_cast<double>(boost::math::iround<double>(hkl[2])));
+      hkl_vectors.push_back(miller_ind);
+      indexed_count++;
+    }
+  }
+
+  if (indexed_count < 3) {
+    throw std::runtime_error(
+        "At least three linearly independent indexed peaks are needed.");
+  }
+  // Get UB using indexed peaks and
+  // lab-Q vectors
+  Matrix<double> UB(3, 3, false);
+  Geometry::IndexingUtils::Optimize_UB(UB, hkl_vectors, peak_q_list);
+  Matrix<double> UBinv(UB);
+  UBinv.Invert();
+  UBinv *= (1.0 / (2.0 * M_PI));
+
+  std::vector<std::pair<double, V3D>> qList;
+  for (size_t i = 0; i < n_peaks; i++) {
+    qList.emplace_back(1., V3D(peaks[i].getQLabFrame()));
+  }
+
+  const bool integrateEdge = getProperty("IntegrateIfOnEdge");
+  if (!integrateEdge) {
+    // This only fails in the unit tests which say that MaskBTP is not
+    // registered
+    try {
+      runMaskDetectors(input_peak_ws, "Tube", "edges");
+      runMaskDetectors(input_peak_ws, "Pixel", "edges");
+    } catch (...) {
+      g_log.error("Can't execute MaskBTP algorithm for this instrument to set "
+                  "edge for IntegrateIfOnEdge option");
+    }
+    calculateE1(
+        input_peak_ws->detectorInfo()); // fill E1Vec for use in detectorQ
+  }
+
+  const bool integrateInHKL = getProperty("IntegrateInHKL");
+  Integrate3DEvents integrator(qList, UBinv, getProperty("RegionRadius"));
+
+  if (eventWS) {
+    // process as EventWorkspace
+    qListFromEventWS(integrator, prog, eventWS, UBinv, integrateInHKL);
+  } else {
+    // process as Workspace2D
+    qListFromHistoWS(integrator, prog, histoWS, UBinv, integrateInHKL);
+  }
+
+  std::vector<std::pair<int, V3D>> weakPeaks, strongPeaks;
+
+  // Compute signal to noise ratio for all peaks
+  int index = 0;
+  for (const auto &item : qList) {
+    const auto center = item.second;
+    IntegrationParameters params = makeIntegrationParameters(center);
+    auto sig2noise = integrator.estimateSignalToNoiseRatio(params, center);
+
+    auto &peak = peak_ws->getPeak(index);
+    peak.setIntensity(0);
+    peak.setSigmaIntensity(0);
+
+    const auto result = std::make_pair(index, center);
+    if (sig2noise < weakPeakThreshold) {
+      g_log.notice() << "Peak " << peak.getHKL() << " with Q = " << center
+                     << " is a weak peak with signal to noise " << sig2noise
+                     << "\n";
+      weakPeaks.push_back(result);
```

Comment:
could this be emplace_back? Could you not create the result object in the push/emplace_back?